src/avro_util.erl

```diff
@@ -295,6 +296,113 @@ resolve_duplicated_refs(Type0) ->
   {Type, _Refs} = resolve_duplicated_refs(Type0, []),
   Type.
 
+%% @doc Check if writer schema is compatible to reader schema.
+-spec is_compatible(avro_type(), avro_type()) -> true | {not_compatible, _, _}.
+is_compatible(Reader, Writer) ->
+  try
+    do_is_compatible_next(Reader, Writer, [], [])
+  catch
+    throw : {not_compatible, RPath, WPath} ->
+      {not_compatible, lists:reverse(RPath), lists:reverse(WPath)};
+    throw : {reader_missing_defalut_value, Path} ->
+      {reader_missing_default_value, lists:reverse(Path)}
+  end.
+
+do_is_compatible_next(Reader, Writer, RPath, WPath) ->
+  NewRPath = [avro:get_type_fullname(Reader) | RPath],
+  NewWPath = [avro:get_type_fullname(Writer) | WPath],
+  do_is_compatible(Reader, Writer, NewRPath, NewWPath).
+
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_RECORD_TYPE(Reader) andalso ?IS_RECORD_TYPE(Writer) ->
+  SameType = avro:is_same_type(Reader, Writer),
+  ReaderTypes = avro_record:get_all_field_data(Reader),
+  WriterTypes = avro_record:get_all_field_data(Writer),
+  SameType andalso
+    lists:all(
+      fun({FieldName, FieldType, Default}) ->
+          case lists:keysearch(FieldName, 1, WriterTypes) of
+            false ->
+              Default =/= ?NO_VALUE orelse
+                erlang:throw({ reader_missing_defalut_value
+                             , [{field, FieldName} |  WPath]
+                             });
+            {value, {_, WriterType, _}} ->
+              FieldDesc = {field, FieldName},
+              do_is_compatible_next(FieldType, WriterType,
+                               [FieldDesc | RPath],
+                               [FieldDesc | WPath])
+          end
+      end,
+      ReaderTypes);
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_ENUM_TYPE(Reader) andalso ?IS_ENUM_TYPE(Writer) ->
+  avro:is_same_type(Reader, Writer) andalso
+    Writer#avro_enum_type.symbols -- Reader#avro_enum_type.symbols == [] orelse
+    erlang:throw({not_compatible, RPath, WPath});
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_FIXED_TYPE(Reader) andalso ?IS_FIXED_TYPE(Writer) ->
+  avro:is_same_type(Reader, Writer)
+    andalso (Reader#avro_fixed_type.size == Writer#avro_fixed_type.size) orelse
+    erlang:throw({not_compatible, RPath, WPath});
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_UNION_TYPE(Reader) andalso ?IS_UNION_TYPE(Writer) ->
```

Comment:
Spec says
> if both are unions:
> The first schema in the reader's union that matches the selected writer's union schema is recursively resolved against it. if none match, an error is signalled.

It sounds like the order of elements doesn't matter and you should check compatibility of each writer schema with each reader schema (like nested loop, pseudo-python-code)
```
for writer in writers:
    found = False
    for reader in readers:
        if test(reader, writer):
            found = True
            break
    if not found:
        return (False, "Can't find reader for writer", writer)
return True
```
and return true if at least one reader is found for each writer.
However, we can leave it as is, so, will have additional requirement to only append new elements to union as the last elements of the list, but then it would be nice to document this requirement.

Reply:
UPD: that's how it's done in Java validator:

https://github.com/apache/avro/blob/cf2f30336efe0ecc3debc7bede86fde6d23f7c79/lang/java/avro/src/main/java/org/apache/avro/SchemaCompatibility.java#L304-L322

looks like

```
do_is_compatible(Reader, Writer, ...) ->
  <..>
  lists:all(
      fun(WriterSubschema) ->
          do_is_compatible(Reader, WriterSubschema)
      end, WriterTypes)
```

Reply:
I think that was from my suggestion.
the spec is all about when a reader member id has been obtained already at run tine when decoding a payload. but we are implementing a check on schema at 'compile' time.

I agree it makes more sense to check if all writer members are compatible with any reader member.